File: sample1/cpp/writer.cpp
/**
 * Protocol buffer data writer in cpp
 *
 * @author Atsushi Sakai
 **/
#include <fstream>
#include <iostream>
#include <string>

#include "addressbook.pb.h"

using namespace std;

int main(void) {
  cout << "Start Protocol buffers writer sample" << endl;
  tutorial::AddressBook address_book;

  tutorial::Person *person1 = address_book.add_people();
  person1->set_id(1234);
  person1->set_name("John Doe");
  person1->set_email("jdoe@example.com");

  tutorial::Person::PhoneNumber *phone = person1->add_phones();
  phone->set_number("555-4321");
  phone->set_type(tutorial::Person::HOME);

  tutorial::Person *person2 = address_book.add_people();
  person2->set_id(4321);
  person2->set_name("Tom Ranger");
  person2->set_email("tranger@example.com");

  tutorial::Person::PhoneNumber *phone2 = person2->add_phones();
  phone2->set_number("555-4322");
  phone2->set_type(tutorial::Person::WORK);

  // Human readable print
  cout << address_book.DebugString() << endl;

  ofstream stream("pbdata_cpp.dat");
  if (!stream.bad()) {
    address_book.SerializeToOstream(&stream);
    stream.close();
  }

  cout << "Done" << endl;
}

File: proto_ver1/java/src/main/java/writer/Writer.java
package writer;

import java.io.FileOutputStream;

import tutorial.Addressbook.AddressBook;
import tutorial.Addressbook.Person;
import tutorial.Addressbook.Person.PhoneNumber;
import static tutorial.Addressbook.Person.PhoneType.*;

public class Writer {

    public static void main(String[] args) throws Exception {
        System.out.println("Protocol buffer writer start!!");

        AddressBook.Builder addressBook = AddressBook.newBuilder();

        Person.Builder person1 = Person.newBuilder();
        person1.setId(1234);
        person1.setName("John Doe");
        person1.setEmail("jdoe@example.com");

        PhoneNumber.Builder phone = PhoneNumber.newBuilder()
                                    .setNumber("555-4321")
                                    .setType(HOME);
        person1.addPhones(phone.build());

        Person.Builder person2 = Person.newBuilder();
        person2.setId(4321);
        person2.setName("Tom Ranger");
        person2.setEmail("tranger@example.com");

        PhoneNumber.Builder phone2 = PhoneNumber.newBuilder()
                                    .setNumber("555-4322")
                                    .setType(WORK);
        person2.addPhones(phone2.build());

        addressBook.addPeople(person1.build());
        addressBook.addPeople(person2.build());


        System.out.println(addressBook); // Human readable print

        FileOutputStream output = new FileOutputStream("pbdata_java.dat");
        addressBook.build().writeTo(output);
        output.close();

        System.out.println("Protocol buffer writer done!!");
        

    }

}
